osu.Server.Queues.ScoreStatisticsProcessor/Commands/Queue/WatchHighScoresCommand.cs

```diff
@@ -136,18 +140,47 @@ public async Task<int> OnExecuteAsync(CancellationToken cancellationToken)
                                              .OrderBy(s => s.score_id)
                                              .ToArray();
 
-                    var pendingProcessing = highScores.FirstOrDefault(s => s.status == 0);
+                    var lastPending = pendingProcessing;
+                    pendingProcessing = highScores.FirstOrDefault(s => s.status == 0);
+
+                    const int sleep = 500;
 
                     if (pendingProcessing != null)
                     {
-                        Console.WriteLine($"Waiting on processing of (score_id: {pendingProcessing.score_id} queue_id: {pendingProcessing.queue_id})");
-                        Thread.Sleep(500);
-                        continue;
+                        // the pending score may have changed, so reset the wait time at this point.
+                        if (pendingProcessing.score_id != lastPending?.score_id)
+                        {
+                            pendingProcessingWaitTime = sleep;
+
+                            Console.WriteLine($"Waiting on processing of (score_id: {pendingProcessing.score_id} queue_id: {pendingProcessing.queue_id})");
+                            Thread.Sleep(sleep);
+                            continue;
+                        }
+
+                        // 5 minute timeout
+                        if (pendingProcessingWaitTime > 300000)
+                        {
+                            Console.WriteLine($"Importing score which was not processed due to timeout (score_id: {pendingProcessing.score_id} queue_id: {pendingProcessing.queue_id})");
+
+                            // only process up to the timed-out entry.
+                            highScores = highScores.Where(s => s.queue_id <= pendingProcessing.score_id).ToArray();
```

Comment:
Uh... isn't this mixing two disparate ID schemes?

```suggestion
                            highScores = highScores.Where(s => s.queue_id <= pendingProcessing.queue_id).ToArray();
```

Also doing things this way means that if legacy processing just dies this is going to throttle everything to a rate of 1 score every 5 minutes. Not sure if that's a problem.

Reply:
The alerts we have setup require higher than that rate, so I think it should be okay. And is probably for the best if more than the odd entry are not being processed for PP.